windows-gaming-driver/src/controller.rs

```diff
@@ -100,66 +107,77 @@ impl Controller {
         // we return false if we didn't notice the GA going down as the timer still
         // exists in that case so it would be a bug to create a second one.
 
-        let ga = self.ga;
-        self.ga = State::Up;
+        let ga = mem::replace(&mut self.ga, State::Up);
         match ga {
             State::Pinging | State::Up => false,
             State::Down => true,
             State::Resuming => {
-                self.set_io_attached(true, false);
+                self.io_attach();
                 true
             }
             State::Suspending => false, // wtf though
         }
     }
 
     pub fn ga_suspending(&mut self) {
-        self.set_io_attached(false, true);
+        self.io_detach();
         self.ga = State::Suspending;
     }
 
-    pub fn set_io_attached(&mut self, state: bool, force: bool) {
-        let go = match self.ga {
-            State::Up | State::Pinging => true,
-            State::Down => force,
+    /// Attaches all configured devices if GA is up
+    pub fn io_attach(&mut self) {
+        match self.ga {
+            State::Down | State::Resuming => (),
             State::Suspending => {
-                assert!(state, "trying to exit from a suspended vm?");
-
                 // make them wake up
                 writemon(&mut self.monitor, &QmpCommand::SystemWakeup);
-
                 // can't enter now - gotta wait for GA to get ready
                 self.ga = State::Resuming;
-                false
-            }
-            State::Resuming => false,
-        };
-
-        if go {
-            match (self.io_attached, state) {
-                (false, true) => {
-                    // attach
-                    for (i, &(vendor, product)) in self.usb_devs.iter().enumerate() {
-                        writemon(&mut self.monitor, &QmpCommand::DeviceAdd {
-                            driver: "usb-host",
-                            id: format!("usb{}", i),
-                            vendorid: vendor,
-                            productid: product,
-                        });
-                    }
-                }
-                (true, false) => {
-                    // detach
-                    for i in 0..self.usb_devs.len() {
-                        writemon(&mut self.monitor, &QmpCommand::DeviceDel { id: format!("usb{}", i) });
-                    }
-                }
-                _ => (),
-            }
-            self.io_attached = state;
+            },
+            State::Up | State::Pinging => self.attach()
+        }
+
+    }
+
+    /// Attaches all configured devices regardless of GA state
+    pub fn io_force_attach(&mut self) {
+        self.attach();
+    }
+
+    /// Detaches all configured devices
+    pub fn io_detach(&mut self) {
+        match self.ga {
+            State::Suspending => panic!("trying to exit from a suspended vm"),
+            _ => self.detach()
```

Comment:
Same here, there's no reason to hide the functionality in a private method.

Also, `assert!(self.ga != State::Suspending)` feels **much** more natural than a match here.